extensions/amp-bind/0.1/bind-impl.js

```diff
@@ -128,235 +124,268 @@ export class Bind {
     });
   }
 
   /**
-   * @param {BindVsyncStateDef} unusedState
+   * Verifies expression results against current DOM state.
+   * @param {Object<string, ./bind-expression.BindExpressionResultDef>} results
    * @private
    */
-  measure_(unusedState) {
-    // TODO(choumx): Validate here or in applyBinding_()?
+  verify_(results) {
+    this.boundElements_.forEach(boundElement => {
+      const {element, bindings} = boundElement;
+
+      bindings.forEach(binding => {
+        const newValue = results[binding.expression];
+        this.verifyBinding_(binding, element, newValue);
+      });
+    });
   }
 
   /**
-   * Either applies or verifies the binding evaluation results in `state`.
-   * @param {BindVsyncStateDef} state
+   * Applies expression results to DOM.
+   * @param {Object<string, ./bind-expression.BindExpressionResultDef>} results
    * @private
    */
-  mutate_(state) {
-    for (let i = 0; i < this.bindings_.length; i++) {
-      const binding = this.bindings_[i];
-      const expression = binding.expression;
-      const result = state.results[expression];
-
-      // Don't apply mutation if the result hasn't changed.
-      if (this.shallowEquals_(result, binding.previousResult)) {
-        continue;
-      } else {
-        binding.previousResult = result;
-      }
+  apply_(results) {
+    this.boundElements_.forEach(boundElement => {
+      const {element, bindings} = boundElement;
+
+      this.resources_.mutateElement(element, () => {
+        const attributeChanges = [];
+        let width, height;
+
+        bindings.forEach(binding => {
+          const newValue = results[binding.expression];
+
+          // Don't apply mutation if the result hasn't changed.
+          if (this.shallowEquals_(newValue, binding.previousResult)) {
+            return;
+          } else {
+            binding.previousResult = newValue;
+          }
 
-      if (state.verifyOnly) {
-        this.verifyBinding_(binding, result);
-      } else {
-        this.applyBinding_(binding, result);
-      }
-    }
+          const attrChange = this.applyBinding_(binding, element, newValue);
+          if (attrChange) {
+            attributeChanges.push(attrChange);
+          }
+
+          switch (binding.property) {
+            case 'width':
+              width = isFiniteNumber(newValue) ? Number(newValue) : width;
+              break;
+            case 'height':
+              height = isFiniteNumber(newValue) ? Number(newValue) : height;
+              break;
+          }
+        });
+
+        if (width !== undefined || height !== undefined) {
+          // TODO(choumx): Add new Resources method for adding change-size
+          // request without scheduling vsync pass since `mutateElement()`
+          // will schedule a pass after a short delay anyways.
+          this.resources_./*OK*/changeSize(element, height, width);
+        }
+
+        if (typeof element.mutatedAttributesCallback === 'function') {
```

Comment:
We can defer to the followup PRs, but some significant questions here are:

1. What if the target is a non-AMP element? E.g. consider `<div hidden>` attribute.
2. What if the target AMP element has not been upgraded yet?